src/main/java/org/hl7/davinci/priorauth/SubscriptionEndpoint.java

```diff
@@ -0,0 +1,156 @@
+                    // Subscription must be rest-hook or websocket....
+                    status = HttpStatus.BAD_REQUEST;
+                    OperationOutcome error = FhirUtils.buildOutcome(IssueSeverity.ERROR, IssueType.INVALID,
+                            INVALID_CHANNEL_TYPE);
+                    formattedData = requestType == RequestType.JSON ? App.getDB().json(error) : App.getDB().xml(error);
+                }
+            } else {
+                // Subscription is required...
+                status = HttpStatus.BAD_REQUEST;
+                OperationOutcome error = FhirUtils.buildOutcome(IssueSeverity.ERROR, IssueType.INVALID,
+                        REQUIRES_SUBSCRIPTION);
+                formattedData = requestType == RequestType.JSON ? App.getDB().json(error) : App.getDB().xml(error);
+            }
+        } catch (Exception e) {
+            // The subscription failed so spectacularly that we need to
+            // catch an exception and send back an error message...
+            status = HttpStatus.BAD_REQUEST;
+            OperationOutcome error = FhirUtils.buildOutcome(IssueSeverity.FATAL, IssueType.STRUCTURE, e.getMessage());
+            formattedData = requestType == RequestType.JSON ? App.getDB().json(error) : App.getDB().xml(error);
+        }
+        MediaType contentType = requestType == RequestType.JSON ? MediaType.APPLICATION_JSON
+                : MediaType.APPLICATION_XML;
+        return ResponseEntity.status(status).contentType(contentType).body(formattedData);
+    }
+
+    private Subscription processSubscription(Subscription subscription) {
+        // Get the criteria details parsed
+        String claimResponseId = "";
+        String patient = "";
+        String status = "";
+        String criteria = subscription.getCriteria();
+        String regex = "identifier=(.*)&patient.identifier=(.*)&status=(.*)";
+
+        Pattern pattern = Pattern.compile(regex);
+        Matcher matcher = pattern.matcher(criteria);
+
+        if (matcher.find() && matcher.groupCount() == 3) {
+            claimResponseId = matcher.group(1);
+            patient = matcher.group(2);
+            status = matcher.group(3);
+        } else {
+            logger.fine("Subscription.criteria: " + criteria);
+            logger.severe("Subcription.criteria is not in the form " + regex);
+            return null;
+        }
+
+        // Add to db
+        String id = UUID.randomUUID().toString();
+        subscription.setId(id);
+        logger.fine("SubscriptionEndpoint::Subscription given uuid " + id);
+        Map<String, Object> dataMap = new HashMap<String, Object>();
+        dataMap.put("id", id);
+        dataMap.put("claimResponseId", claimResponseId);
+        dataMap.put("patient", patient);
+        dataMap.put("status", status);
+        dataMap.put("resource", subscription);
+        if (App.getDB().write(Database.SUBSCRIPTION, dataMap))
+            return subscription;
+        else
+            return null;
```

Comment:
Is it better to generate and throw an exception here, or let the caller do it?

Reply:
The rest of the error handling is dealt with in the caller because that is where the actual response is being created. This method only returns a subscription so I think it's easier and cleaner to have the caller process the exception and then send an appropriate response back to the user